File: 39 Combination Sum/Solution.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(candidates);
        recurse(candidates, target, 0, new ArrayList<Integer>(), res);
        return res;
    }
    public void recurse(int[] candidates, int target, int start, List<Integer> cur, List<List<Integer>> res) {
        if (target < 0)
            return;
        if (target == 0) {
            res.add(new ArrayList<Integer>(cur));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            cur.add(candidates[i]);
            recurse(candidates, target-candidates[i], i, cur, res);
            cur.remove(cur.size()-1);
        }
    }
}

File: 39 Combination Sum/Solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());
        vector<int> vec;
        recurse(candidates, target, 0, vec, res);
        return res;
    }
    void recurse(vector<int>& candidates, int target, int index, vector<int> vec, vector<vector<int>> res) {
        if (target < 0)
            return;
        if (target == 0) {
            res.push_back(vec);
            return;
        }

        for (int i = index; i < candidates.size(); i++) {
            vec.push_back(candidates[i]);
            recurse(candidates, target - candidates[i], index, vec, res);
            vec.pop_back();
        }
    }
};
